packages/core/src/shared/utilities/timeoutUtils.ts

```diff
@@ -267,6 +273,17 @@ export async function waitUntil<T>(fn: () => Promise<T>, options: WaitUntilOptio
     let elapsed: number = 0
     let remaining = opt.timeout
 
+    // Internal helper to determine if we should retry
+    function shouldRetry(error: Error | undefined): boolean {
+        if (error === undefined) {
+            return typeof opt.retryOnFail === 'boolean' ? opt.retryOnFail : true
+        }
+        if (typeof opt.retryOnFail === 'function') {
```

Comment:
Can we make it always be a function? That should simplify some logic, eliminate an overload, and eliminate some tests.

It's trivial for calls to pass `() => true` if they want a "constant".